## Problem #174: Hollow Square Laminae II
[Link to Problem](https://projecteuler.net/problem=174)

### Problem Description

We shall define a square lamina to be a square outline with a square "hole" so that the shape possesses vertical and horizontal symmetry.

Given eight tiles it is possible to form a lamina in only one way: $3 \times 3$ square with a $1 \times 1$ hole in the middle. However, using thirty-two tiles it is possible to form two distinct laminae.

![example](https://projecteuler.net/resources/images/0173_square_laminas.gif?1678992055)

If $t$ represents the number of tiles used, we shall say that $t = 8$ is type $L(1)$ and $t = 32$ is type $L(2)$.

Let $N(n)$ be the number of $t \le 1000000$ such that $t$ is type $L(n)$; for example, $N(15) = 832$.

What is $\sum\limits_{n = 1}^{10} N(n)$?

### Approach

Let $a$ = the side of the outer square and $b$ = the side of the inner square.

$$

t = a^2 - b^2 = (a - b)(a + b)

$$

Also, $a$ and $b$ should have the same parity.

Therefore we can iterate $a$ from 1 to $10^6$ then iterate $b$ from $a - 2$ to $1$ and once $t$ becomes too big we stop iterating $b$ and try the next $a$.

In [1]:
limit = 1_000_000

ways = [0] * (limit + 1)

for a in range(3, limit + 1):
    for b in range(a - 2,  0, -2):
        t = a * a - b * b
        if t > limit:
            break
        ways[t] += 1

answer = 0

for t in range(1, limit + 1):
    if 1 <= ways[t] <= 10:
        answer += 1

print(answer)

209566


###### Result: **209566** | Execution time: 1s

### Complexity analysis

Time complexity: $O(limit * \log{limit})$ ???

##### Tags: #math, #squares, #number-theory